File: android/src/main/kotlin/com/degastonapps/flutter_tuner/TunerNew.kt
package com.degastonapps.flutter_tuner

import android.annotation.SuppressLint
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.util.Log
import kotlin.math.abs
import kotlin.math.max


class TunerNew(val callback: (Double) -> Unit) {

    companion object {
        const val TAG = "TUNER"
        const val SAMPLE_RATE = 16000
        const val RECORDER_CHANNELS = AudioFormat.CHANNEL_IN_MONO
        const val RECORDER_AUDIO_ENCODING = AudioFormat.ENCODING_PCM_16BIT

        const val ERROR_FREQUENCY = -1.0
        const val TOO_QUIET_FREQ = -2.0;

        // Want to record at least 0.25 seconds data for accurate output of low pitch instruments
        const val BUFFER_SIZE = 4096

        // 2 bytes in 16bit PCM format
        const val BYTES_PER_ELEMENT = 2

        // Tuner specific values:
        const val ATTENUATION = 0.80

        // TODO: Fix the noise levels.
        // Sometimes it picks up frequencies around ~60Hz or ~120Hz
        const val NOISE_SENSE_LOW = 5;
        const val NOISE_SENSE_MED = 10;
        const val NOISE_SENSE_HIGH = 15;
        const val NOISE_SENSE_VERY_HIGH = 40;
    }

    private var isRecording = false
    private var isTuning = false
    private var recorder: AudioRecord? = null
    private var recordingThread: Thread? = null


    private var prevFreq = ERROR_FREQUENCY
    private var prevFreq2 = ERROR_FREQUENCY

    private fun debounceFrequency(freq: Double): Double {
        val average = (prevFreq + prevFreq2 + freq) / 3.0
        val diffPrev = abs(prevFreq - average)
        val diffPrev2 = abs(prevFreq2 - average)
        val diffFreq = abs(freq - average)

        val freqToReturn = when {
            diffPrev <= diffPrev2 && diffPrev <= diffFreq -> prevFreq
            diffPrev2 <= diffPrev && diffPrev2 <= diffFreq -> prevFreq2
            else -> freq
        }

        prevFreq2 = prevFreq
        prevFreq = freq
        return freqToReturn
    }

    private fun collectData(): ShortArray {
        // Write the input audio in byte array
        val sData = ShortArray(BUFFER_SIZE)

        // Get the sound input from microphone
        val shortsRead = recorder!!.read(sData, 0, BUFFER_SIZE)

        if (shortsRead < 0) {
            Log.e(TAG, "collectData: Couldn't read from recorder! shortsRead=$shortsRead")
        }

        return sData
    }

    /**
     * Main frequency estimation method.
     */
    private fun findFrequency(sigIn: DoubleArray, noiseFloor: Double = NoiseLevels.NOISE_SENSE_VERY_HIGH): Double {
        val sig = sigIn.copyOf()
        if (sig.size < BUFFER_SIZE) {
            Log.e(TAG, "Not enough samples! Expecting > $BUFFER_SIZE. Got ${sig.size}")
            return ERROR_FREQUENCY
        }

        // Split into overlapping windows (Welch-style)
        val factor = 4 // must be multiple of 4 as in Dart
        val windowLength = sig.size / factor
        var startIndex = 0
        var windowsCalculated = 0

        // powers array length = windowLength/2 + 1
        val powers = DoubleArray(windowLength / 2 + 1)

        val ham = hamming(windowLength)

        // Perform factor*2 - 1 windows (overlap by half-window)
        while (windowsCalculated != factor * 2 - 1) {
            val windowed = DoubleArray(windowLength)
            for (i in 0 until windowLength) {
                windowed[i] = sig[startIndex + i] * ham[i]
            }

            val f = rfft(windowed) // Real FFT result
            val fAbs = arrayComplexAbs(f) // DoubleArray

            // powers += fAbs element-wise
            for (i in powers.indices) {
                powers[i] += fAbs.getOrElse(i) { 0.0 }
            }

            startIndex += windowLength / 2
            windowsCalculated++
        }

        var foundSignal = false
        val avgPower = mean(powers)

        // Find top NUM_PEAKS peaks
        val peaks = mutableListOf<Int>()
        val NUM_PEAKS = 5
        val START_FFT_INDEX = 2
        for (i in 0 until NUM_PEAKS) {
            var tmpMax = -1.0
            var tmpIdx = -1
            for (j in START_FFT_INDEX until powers.size) {
                if (powers[j] > tmpMax && !peaks.contains(j)) {
                    tmpIdx = j
                    tmpMax = powers[j]
                    if (tmpMax > avgPower * noiseFloor) {
                        foundSignal = true
                    }
                }
            }
            peaks.add(tmpIdx)
        }

        if (!foundSignal) {
            Log.d(TAG, "TOO QUIET! avgPower=$avgPower")
            return TOO_QUIET_FREQ
        }

        val maxIndex = scorePowers(powers, peaks)

        // Parabolic interpolation for sub-bin accuracy
        val logPowers = arrayLog(powers)
        val par = parabolic(logPowers, peaks[maxIndex])
        val trueI = par[0]

        val newFreq = SAMPLE_RATE * trueI / windowLength
        if (newFreq < 0.0) return ERROR_FREQUENCY

        return newFreq
    }

    private fun scorePowers(powers: DoubleArray, peaks: List<Int>): Int {
        // Harmonic Product Spectrum-inspired scoring
        val sums = mutableListOf<Double>()
        val factors = listOf(2.0, 3.0, 4.0)

        for (peak in peaks) {
            var sum = 0.0
            if (peak <= 0) {
                sums.add(sum)
                continue
            }

            for (fFactor in factors) {
                var newIndex = (peak * fFactor).toInt()
                if (newIndex < 2) newIndex = 2
                if (newIndex + 3 > powers.size) newIndex = powers.size - 3

                // include neighbors
//                sum += powers.getOrElse(newIndex - 2) { 0.0 } * powers[peak]
                sum += powers.getOrElse(newIndex - 1) { 0.0 } * powers[peak]
                sum += powers.getOrElse(newIndex) { 0.0 } * powers[peak]
                sum += powers.getOrElse(newIndex + 1) { 0.0 } * powers[peak]
//                sum += powers.getOrElse(newIndex + 2) { 0.0 } * powers[peak]
            }
            sums.add(sum)
        }

        var maxIndex = 0
        for (i in 1 until sums.size) {
            if (sums[i] > sums[maxIndex]) {
                maxIndex = i
            }
        }

        return maxIndex
    }


    /**
     * Check for permissions BEFORE calling this function
     */
    @SuppressLint("MissingPermission")
    fun start() {
        if (isRecording) return // Already running, ignore

        Log.d(TAG, "start: Starting tuner")

        recorder = AudioRecord(
            MediaRecorder.AudioSource.MIC,
            SAMPLE_RATE,
            RECORDER_CHANNELS,
            RECORDER_AUDIO_ENCODING,
            BUFFER_SIZE * BYTES_PER_ELEMENT
        )

        recorder?.startRecording()
        isRecording = true

        recordingThread = Thread({
            while (isRecording) {
                try {
                    val sData = collectData()
                    val doubleData = sData.map { it.toDouble() }.toDoubleArray()
                    val freq = findFrequency(doubleData)
                    callback(freq)
                } catch (e: Exception) {
                    Log.e(TAG, "Error: ${e.message}\n Stack: ${e.stackTraceToString()}")
                    callback(ERROR_FREQUENCY)
                }
            }
        }, "recordingThread")

        recordingThread?.start()
    }


    fun stop() {
        if (!isRecording) return // Already stopped, ignore

        Log.d(TAG, "stop: Stopping tuner")
        isTuning = false
        isRecording = false

        try {
            recorder?.stop()
            recorder?.release()
        } catch (e: Exception) {
            Log.e(TAG, "stop: Error stopping recorder", e)
        }

        recorder = null

        try {
            recordingThread?.join()
        } catch (e: InterruptedException) {
            Log.e(TAG, "stop: Thread join interrupted", e)
        }

        recordingThread = null // Clear so a new one can be created
    }

}

/* ------------------------------------------------------------------------
   DSP helper placeholders. Implement these with a fast native/DSP library
   (JTransforms, FFTW wrapper, kissfft, or your own SIMD code).
   ------------------------------------------------------------------------ */
data class Complex(val re: Double, val im: Double) {
    operator fun plus(other: Complex) = Complex(re + other.re, im + other.im)
    operator fun minus(other: Complex) = Complex(re - other.re, im - other.im)
    operator fun times(other: Complex) =
        Complex(re * other.re - im * other.im, re * other.im + im * other.re)
    fun scale(f: Double) = Complex(re * f, im * f)
}

/**
 * Window function.
 * Returns length-sized Hamming window.
 */
fun hamming(length: Int): DoubleArray {
    val out = DoubleArray(length)
    for (n in 0 until length) {
        out[n] = 0.54 - 0.46 * kotlin.math.cos(2.0 * Math.PI * n / (length - 1))
    }
    return out
}


/**
 * Recursive Cooley–Tukey FFT
 */
private fun fft(input: Array<Complex>): Array<Complex> {
    val n = input.size
    if (n == 1) return arrayOf(input[0])

    if (n % 2 != 0) throw IllegalArgumentException("FFT input length must be power of 2")

    val even = fft(Array(n / 2) { input[2 * it] })
    val odd = fft(Array(n / 2) { input[2 * it + 1] })

    val out = Array(n) { Complex(0.0, 0.0) }
    for (k in 0 until n / 2) {
        val angle = -2.0 * Math.PI * k / n
        val wk = Complex(Math.cos(angle), Math.sin(angle)) * odd[k]
        out[k] = even[k] + wk
        out[k + n / 2] = even[k] - wk
    }
    return out
}

/**
 * Real FFT. Returns bins of length = windowLength/2 + 1
 */
fun rfft(windowed: DoubleArray): Array<Complex> {
    val n = windowed.size
    if (n and (n - 1) != 0) throw IllegalArgumentException("Input length must be power of 2")

    val complexInput = Array(n) { Complex(windowed[it], 0.0) }
    val spectrum = fft(complexInput)

    // Only keep positive frequencies [0..N/2]
    return spectrum.copyOfRange(0, n / 2 + 1)
}


/**
 * Compute magnitudes of complex array
 */
fun arrayComplexAbs(c: Array<Complex>): DoubleArray {
    return DoubleArray(c.size) { i ->
        kotlin.math.hypot(c[i].re, c[i].im)
    }
}

/**
 * Mean of an array
 */
fun mean(arr: DoubleArray): Double {
    if (arr.isEmpty()) return 0.0
    var sum = 0.0
    for (v in arr) sum += v
    return sum / arr.size
}

/**
 * Natural log of array elements. Returns new DoubleArray.
 */
fun arrayLog(arr: DoubleArray): DoubleArray {
    return DoubleArray(arr.size) { i ->
        kotlin.math.ln(max(arr[i], Double.MIN_VALUE))
    }
}

/**
 * Parabolic interpolation around index `i` on array `y` (log-powers).
 * Returns DoubleArray where index 0 = true_i (interpolated peak index),
 * index 1 = interpolated value (optional).
 *
 * Implementation expects `i` to be a valid index within y.
 * TODO: you can replace with a version that returns a richer structure.
 */
fun parabolic(y: DoubleArray, i: Int): DoubleArray {
    // i should have neighbors; clamp to safe range
    val idx = i.coerceIn(1, y.size - 2)
    val alpha = y[idx - 1]
    val beta = y[idx]
    val gamma = y[idx + 1]
    val p = 0.5 * (alpha - gamma) / (alpha - 2.0 * beta + gamma)
    val trueI = idx + p
    val value = beta - 0.25 * (alpha - gamma) * p
    return doubleArrayOf(trueI, value)
}

/* ------------------------------------------------------------------------
   com.degastonapps.flutter_tuner.NoiseLevels object mirrors your dart noise_levels import.
   Replace values with your project's constants.
   ------------------------------------------------------------------------ */
object NoiseLevels {
    // Example value. Set to the constant you use in Dart.
    const val NOISE_SENSE_VERY_HIGH: Double = 25.0
}

File: ios/Classes/TunerNew.swift
import Foundation
import AVFoundation
import Accelerate

class TunerNew {

    // MARK: - Constants

    private static let sampleRate = 16000.0
    private static let bufferSize = 4096

    private static let errorFrequency = -1.0
    private static let tooQuietFreq = -2.0

    // This value is taken from the Android implementation's NoiseLevels object.
    private static let noiseSenseVeryHigh = 25.0

    // MARK: - Properties

    private let callback: (Double) -> Void
    private var audioEngine = AVAudioEngine()
    private var isRecording = false

    // MARK: - Initialization

    init(callback: @escaping (Double) -> Void) {
        self.callback = callback
    }

    // MARK: - Public Methods

    func start() throws {
        guard !isRecording else { return }

        // 1. Configure the audio session.
        let session = AVAudioSession.sharedInstance()
        try session.setCategory(.playAndRecord, mode: .measurement, options: .duckOthers)
        try session.setPreferredSampleRate(TunerNew.sampleRate)
        try session.setActive(true)

        // 2. Get the input node and its actual output format.
        let inputNode = audioEngine.inputNode
        let recordingFormat = inputNode.outputFormat(forBus: 0)

        // 3. Install a tap to process the audio buffer.
        inputNode.installTap(onBus: 0, bufferSize: AVAudioFrameCount(TunerNew.bufferSize), format: recordingFormat) { (buffer, time) in
            guard self.isRecording, let channelData = buffer.floatChannelData else { return }

            let frameLength = Int(buffer.frameLength)
            let samples = Array(UnsafeBufferPointer(start: channelData[0], count: frameLength))

            let doubleData = samples.map { Double($0) }
            let frequency = self.findFrequency(sigIn: doubleData)

            self.callback(frequency)
        }

        // 4. Prepare and start the audio engine.
        audioEngine.prepare()
        try audioEngine.start()
        isRecording = true
    }

    func stop() {
        guard isRecording else { return }
        isRecording = false
        audioEngine.stop()
        audioEngine.inputNode.removeTap(onBus: 0)

        do {
            try AVAudioSession.sharedInstance().setActive(false)
        } catch {
            print("Error deactivating audio session: \(error.localizedDescription)")
        }
    }

    // MARK: - Private DSP Methods

    private func findFrequency(sigIn: [Double], noiseFloor: Double = TunerNew.noiseSenseVeryHigh) -> Double {
        if sigIn.count < TunerNew.bufferSize {
            return TunerNew.errorFrequency
        }

        let factor = 4
        let windowLength = sigIn.count / factor
        var startIndex = 0
        var windowsCalculated = 0

        var powers = [Double](repeating: 0.0, count: windowLength / 2 + 1)
        let ham = hamming(length: windowLength)

        while windowsCalculated != factor * 2 - 1 {
            let windowed = (0..<windowLength).map { sigIn[startIndex + $0] * ham[$0] }

            if let f = rfft(windowed), f.count > 0 {
                let fAbs = arrayComplexAbs(f)
                for i in 0..<min(powers.count, fAbs.count) {
                    powers[i] += fAbs[i]
                }
            }

            startIndex += windowLength / 2
            windowsCalculated += 1
        }

        var foundSignal = false
        let avgPower = mean(arr: powers)

        var peaks: [Int] = []
        let numPeaks = 5
        let startFFTIndex = 2
        for _ in 0..<numPeaks {
            var tmpMax = -1.0
            var tmpIdx = -1
            for j in startFFTIndex..<powers.count {
                if powers[j] > tmpMax && !peaks.contains(j) {
                    tmpIdx = j
                    tmpMax = powers[j]
                    if tmpMax > avgPower * noiseFloor {
                        foundSignal = true
                    }
                }
            }
            if tmpIdx != -1 {
                peaks.append(tmpIdx)
            }
        }

        if !foundSignal || peaks.isEmpty {
            return TunerNew.tooQuietFreq
        }

        let maxIndex = scorePowers(powers: powers, peaks: peaks)

        let logPowers = arrayLog(arr: powers)
        let par = parabolic(y: logPowers, i: peaks[maxIndex])
        let trueI = par[0]

        let actualSampleRate = audioEngine.inputNode.outputFormat(forBus: 0).sampleRate
        let newFreq = actualSampleRate * trueI / Double(windowLength)
        return newFreq < 0.0 ? TunerNew.errorFrequency : newFreq
    }

    private func scorePowers(powers: [Double], peaks: [Int]) -> Int {
        var sums: [Double] = []
        let factors = [2.0, 3.0, 4.0]

        for peak in peaks {
            var sum = 0.0
            if peak <= 0 {
                sums.append(sum)
                continue
            }

            for fFactor in factors {
                var newIndex = Int(Double(peak) * fFactor)
                if newIndex < 2 { newIndex = 2 }
                if newIndex + 2 >= powers.count { newIndex = powers.count - 3 }

                sum += powers[newIndex - 1] * powers[peak]
                sum += powers[newIndex] * powers[peak]
                sum += powers[newIndex + 1] * powers[peak]
            }
            sums.append(sum)
        }

        var maxIndex = 0
        if let maxVal = sums.max(), let firstIndex = sums.firstIndex(of: maxVal) {
            maxIndex = firstIndex
        }

        return maxIndex
    }

    // MARK: - DSP Helper Functions

    private func hamming(length: Int) -> [Double] {
        return vDSP.window(ofType: Double.self, usingSequence: .hamming, count: length, isHalfWindow: false)
    }

    private func rfft(_ x: [Double]) -> [DSPDoubleComplex]? {
        let n = x.count
        guard n > 0, (n & (n - 1)) == 0 else {
            return nil
        }

        let log2n = vDSP_Length(log2(Double(n)))
        guard let fftSetup = vDSP_create_fftsetupD(log2n, FFTRadix(kFFTRadix2)) else {
            return nil
        }
        defer { vDSP_destroy_fftsetupD(fftSetup) }

        var real = [Double](repeating: 0, count: n / 2)
        var imag = [Double](repeating: 0, count: n / 2)

        x.withUnsafeBufferPointer { bufferPtr in
            var complex = DSPDoubleSplitComplex(realp: &real, imagp: &imag)
            vDSP_ctozD(bufferPtr.baseAddress!.withMemoryRebound(to: DSPDoubleComplex.self, capacity: x.count/2) { $0 }, 2, &complex, 1, vDSP_Length(n / 2))
        }

        var complex = DSPDoubleSplitComplex(realp: &real, imagp: &imag)
        vDSP_fft_zripD(fftSetup, &complex, 1, log2n, FFTDirection(kFFTDirection_Forward))

        let dc = complex.realp[0]
        let nyquist = complex.imagp[0]

        var result: [DSPDoubleComplex] = [DSPDoubleComplex(real: dc, imag: 0.0)]
        for i in 1..<(n/2) {
            result.append(DSPDoubleComplex(real: complex.realp[i], imag: complex.imagp[i]))
        }
        result.append(DSPDoubleComplex(real: nyquist, imag: 0.0))

        return result
    }

    private func arrayComplexAbs(_ c: [DSPDoubleComplex]) -> [Double] {
        var real = c.map { $0.real }
        var imag = c.map { $0.imag }
        var mags = [Double](repeating: 0.0, count: c.count)

        var splitComplex = DSPDoubleSplitComplex(realp: &real, imagp: &imag)
        vDSP_zvabsD(&splitComplex, 1, &mags, 1, vDSP_Length(c.count))

        return mags
    }

    private func mean(arr: [Double]) -> Double {
        if arr.isEmpty { return 0.0 }
        return vDSP.mean(arr)
    }

    private func arrayLog(arr: [Double]) -> [Double] {
        var results = [Double](repeating: 0.0, count: arr.count)
        var mutableArr = arr.map { max($0, Double.leastNormalMagnitude) }
        vvlog(&results, &mutableArr, [Int32(arr.count)])
        return results
    }

    private func parabolic(y: [Double], i: Int) -> [Double] {
        let idx = max(1, min(i, y.count - 2))
        let alpha = y[idx - 1]
        let beta = y[idx]
        let gamma = y[idx + 1]

        let denominator = alpha - 2.0 * beta + gamma
        if abs(denominator) < 1e-6 {
            return [Double(idx), beta]
        }

        let p = 0.5 * (alpha - gamma) / denominator
        let trueI = Double(idx) + p
        let value = beta - 0.25 * (alpha - gamma) * p

        return [trueI, value]
    }
}

enum TunerError: Error {
    case invalidFormat
    case audioSessionError(String)
}